doc/sphinx/proof-engine/ltac.rst

```diff
@@ -52,127 +66,98 @@ the variable :token:`ident` matches any complex expression with (possible)
 dependencies in the variables :n:`@ident__i` and returns a functional term
 of the form :n:`fun @ident__1 … ident__n => @term`.
 
-The main entry of the grammar is :n:`@ltac_expr`. This language is used in proof
-mode but it can also be used in toplevel definitions as shown below.
+.. note::
+
+   - If you want :n:`@tactic__2; @tactic__3` to be fully applied to the first
+     subgoal generated by :n:`@tactic__1` before applying it to the other
+     subgoals, then you should write:
+
+     - :n:`@tactic__1; [> @tactic__2; @tactic__3 .. ]` rather than
+
+     - :n:`@tactic__1; (@tactic__2; @tactic__3)`.
+
+   .. todo PR I tried to improve the wording for the following bullet,
+      but I'm not quite sure what it refers to.  Perhaps a fool's errand.
+      Does this issue still come up, and if so, exactly which productions?
```

Comment:
Does this also apply to qualids that appear in binder_tactics or simple_tactics?  If it's not specific to tactic_arg_compat, it should be explained more generally.

Also seems to me that `tactic_arg` should have a different name and that `tactic_arg_compat` should be called `tactic_arg`.  Any ideas?

Reply:
>Does this also apply to qualids that appear in binder_tactics or simple_tactics?

I don't think so.

> Also seems to me that `tactic_arg` should have a different name and that `tactic_arg_compat` should be called `tactic_arg`. Any ideas?

I agree and I already did a suggestion about this: rename the current `tactic_arg` into `ltac_value` and the current `tactic_arg_compat` into `tactic_arg`.

Reply:
How about this?

```
Ltac X := idtac "hi".
Ltac X2 := let Y := X in Y.
X2.   (* output:  "hi" *)
```

But this doesn't work:

```
Parameter P : nat -> Prop.
Goal P 1 /\ P 2 /\ P 3.
repeat split.
Ltac one := 1.
Ltac two := 2.
swap one two.  (* Unbound variable one. *)
```

OTOH, this works.

```
Parameter P : nat -> Prop.
Goal P 1 /\ P 2 /\ P 3.
repeat split.
ltac: (let one := 1 with two := 2 in swap one two).
```

We oughta explain where and how substitution applies.  And name spaces (Ltac tactic names vs let symbols in ltac_exprs).

> rename the current `tactic_arg` into `ltac_value`

I think I lean toward `value_tactic` as we're referring to them as a flavor of tactics.

Reply:
> ```
> Ltac X := idtac "hi".
> Ltac X2 := let Y := X in Y.
> X2.   (* output:  "hi" *)
> ```

Yes, if you use a qualid as a tactic (and not in argument position), it should always be interpreted as a tactic. So the note that we were discussing does not apply.

> But this doesn't work:
> 
> ```
> Parameter P : nat -> Prop.
> Goal P 1 /\ P 2 /\ P 3.
> repeat split.
> Ltac one := 1.
> Ltac two := 2.
> swap one two.  (* Unbound variable one. *)
> ```

That's because `swap` expects a `int_or_var`, not a `tactic_arg_compat`.

> OTOH, this works.
> 
> ```
> Parameter P : nat -> Prop.
> Goal P 1 /\ P 2 /\ P 3.
> repeat split.
> ltac: (let one := 1 with two := 2 in swap one two).
> ```

In this case, the outer `ltac:( )` is useless. However, I do not understand how this can be parsed based on the rules that you show in the doc. This `ltac:( )` syntax should only be accepted in `tactic_arg_compat`, which cannot appear in such position AFAICT.

> We oughta explain where and how substitution applies. And name spaces (Ltac tactic names vs let symbols in ltac_exprs).

Sure lots of things to explain (out of scope for this PR though).

> > rename the current `tactic_arg` into `ltac_value`
> 
> I think I lean toward `value_tactic` as we're referring to them as a flavor of tactics.

Sounds good.

Reply:
> In this case, the outer `ltac:( )` is useless. However, I do not understand how this can be parsed based on the rules that you show in the doc.

It's part of `term`:

```
term0: [
| term_ltac

term_ltac: [
| "ltac" ":" "(" ltac_expr ")"
```

Also, the `let` is an ltac let, not a term let; `with` is only supported in ltac let.  This also works:
`all: let one := 1 with two := 2 in swap one two.`

Reply:
My original wording was incorrect.  It's not a grammatical ambiguity.  Does this look reasonably accurate?

```
   - :token:`qualid`\s within |Ltac| expressions are used to represent both terms and |Ltac| variables.
     If the :token:`qualid` corresponds to
     an |Ltac| variable or tactic name, |Ltac| substitutes the value before applying
     the expression.  Generally it's best to choose distinctive names |Ltac| variable
     that won't clash with term names.  You can use :n:`ltac:(@qualid)` or :n:`(@qualid)`
     to control whether a :token:`qualid` is interpreted as, respectively, an |Ltac|
     variable or a term.

```

Reply:
> > In this case, the outer `ltac:( )` is useless. However, I do not understand how this can be parsed based on the rules that you show in the doc.
> 
> It's part of `term`:
> 
> ```
> term0: [
> | term_ltac
> 
> term_ltac: [
> | "ltac" ":" "(" ltac_expr ")"
> ```

I'm well aware of that. What puzzles me is that a term is accepted in this position.

> ```
> 
> ```
> 
> 
> Also, the `let` is an ltac let, not a term let; `with` is only supported in ltac let.  This also works:

Yes, I'm well aware that this is an Ltac `let`. `with` is likely mostly intended for use in mutually-recursive definitions.

> My original wording was incorrect. It's not a grammatical ambiguity. Does this look reasonably accurate?
> 
> ```
>    - :token:`qualid`\s within |Ltac| expressions are used to represent both terms and |Ltac| variables.
>      If the :token:`qualid` corresponds to
>      an |Ltac| variable or tactic name, |Ltac| substitutes the value before applying
>      the expression.  Generally it's best to choose distinctive names |Ltac| variable
>      that won't clash with term names.  You can use :n:`ltac:(@qualid)` or :n:`(@qualid)`
>      to control whether a :token:`qualid` is interpreted as, respectively, an |Ltac|
>      variable or a term.
> ```

LGTM (apart from a missing word after "distinctive names").

Reply:
> What puzzles me is that a term is accepted in this position.

See the "todo PR" starting with "This is defined programmatically rather than in the mlgs."  The production comes from there.  And please respond to that comment--I guess we should add those to the grrammar.

> I'm well aware of that.

I'm still learning so much of this, I'm apt to state things that are obvious to you from time to time.  And get confused, make mistakes, etc.

Reply:
To be a bit more explicit @Zimmi48 , the reason this is accepted is because it is included in the grammar for `tactic_arg`, which seems largely undocumented in the new grammar, as @jfehrle points out.